extensions/amp-social-share/0.2/amp-social-share.js

```diff
@@ -0,0 +1,176 @@
+   * @return {!JsonObject|undefined}
+   */
+  getTypeConfigOrUndefined_(type, viewer, platform) {
+    userAssert(
+      !/\s/.test(type),
+      'Space characters are not allowed in type attribute value. %s',
+      this.element
+    );
+    if (type === 'system') {
+      // navigator.share unavailable
+      if (!this.systemShareSupported_(viewer, platform)) {
+        return;
+      }
+    } else {
+      // system share wants to be unique
+      const systemOnly =
+        this.systemShareSupported_(viewer, platform) &&
+        !!this.element.ownerDocument.querySelector(
+          'amp-social-share[type=system][data-mode=replace]'
+        );
+      if (systemOnly) {
+        return;
+      }
+    }
+    const typeConfig = getSocialConfig(type) || dict();
+    if (typeConfig['obsolete']) {
+      user().warn(TAG, `Skipping obsolete share button ${type}`);
+      return;
+    }
+    return typeConfig;
+  }
+
+  /**
+   * @private
+   * @param {!../../../src/service/viewer-interface.ViewerInterface} viewer
+   * @param {!../../../src/service/platform-impl.Platform} platform
+   * @return {boolean}
+   */
+  systemShareSupported_(viewer, platform) {
+    // Chrome exports navigator.share in WebView but does not implement it.
+    // See https://bugs.chromium.org/p/chromium/issues/detail?id=765923
+    const isChromeWebview = viewer.isWebviewEmbedded() && platform.isChrome();
+
+    return 'share' in navigator && !isChromeWebview;
+  }
+
+  /**
+   * Sets 'href' and 'target' in the AmpContext based on AMP-specific conditions.
+   * These needs to be passed to the context as opposed to as properties in `init()`
+   * because their values are resolved asynchronously.
+   *
+   * @private
+   * @param {!JsonObject} typeConfig
+   * @param {!../../../src/service/platform-impl.Platform} platform
+   */
+  setHrefAndTargetContext_(typeConfig, platform) {
+    const shareEndpoint = user().assertString(
+      this.element.getAttribute('data-share-endpoint') ||
+        typeConfig['shareEndpoint'],
+      'The data-share-endpoint attribute is required. %s'
+    );
+    const urlParams = getDataParamsFromAttributes(this.element);
+    Object.assign(urlParams, typeConfig['defaultParams']);
+    const hrefWithVars = addParamsToUrl(shareEndpoint, urlParams);
+    const urlReplacements = Services.urlReplacementsForDoc(this.element);
+    const bindingVars = typeConfig['bindings'];
+    const bindings = {};
+    if (bindingVars) {
+      bindingVars.forEach(name => {
+        const bindingName = name.toUpperCase();
+        bindings[bindingName] = urlParams[name];
+      });
+    }
+
+    urlReplacements.expandUrlAsync(hrefWithVars, bindings).then(result => {
+      let href = result;
+      // mailto:, sms: protocols breaks when opened in _blank on iOS Safari
+      const {protocol} = Services.urlForDoc(this.element).parse(href);
+      const isMailTo = protocol === 'mailto:';
+      const isSms = protocol === 'sms:';
+      this.context_['target'] =
```

Comment:
Why are these hoisted on the context? `AmpContext` specifically should have nothing but `renderable` and `playable`. From what I see this does not belong in a context, which would only be needed for cross-element communication.

Reply:
The `href` and `target` rely on `expandUrlAsync` to resolve so cannot be done in `init`, or am I missing something here?

Reply:
I specially meant `this.context_[] = `. The `context_`, which is `AmpContext`, as well as any other React context are used exclusively for cross-component communication. This is not the case here. And overall this seems like a lot of logic might be happening outside the component. Shouldn't component itself be responsible for the correct target calculation?

Reply:
I agree with Dima. This should be a prop, not a context.

Reply:
To clarify, you recommend moving this logic to the React component to resolve itself rather than taking `href` and `target` params? Would that require re-implementing URL replacement services for Preact mode?

Reply:
I removed use of `AmpContext` but there is still the open question of making something like the url replacements service available in Preact mode. Do we have existing tools to resolve the `href` and `target` params in the Component?

Reply:
No, I'm saying you as an Bento element author are not allowed to assign to `context`. You _can_ assign this as a prop that is directly passed to the component. You just need to build that part.

We should not have the URL replacements in Preact. That's best left to whoever is passing in the props to control.

Reply:
From what I can tell, the main parameter on how `href` and `target` are calculated is `type`, right? What do we use url-replacements for then? For canonical URL? Anything else?

Reply:
I replaced `expandUrlAsync` with `expandUrlSync` to be able to return the `href` and `target` as props through `init`

Reply:
Discussed offline -- the AMP element triggers render on the Component once these values are resolved.